pkg/util/loop/loop.go

```diff
@@ -99,106 +113,180 @@ func (loop *Device) AttachFromFile(image *os.File, mode int, number *int) error
-				sylog.Debugf("Could not get loop device %d status: %s", device, err)
-				continue
-			}
-			// there is no associated image with loop device, save indice so second loop
-			// iteration will start from this device
-			if status.Inode == 0 && freeDevice == -1 {
-				freeDevice = device
-				syscall.Close(loopFd)
-				continue
-			}
-			if status.Inode == imageIno && status.Device == imageDev &&
-				status.Flags&FlagsReadOnly == loop.Info.Flags&FlagsReadOnly &&
-				status.Offset == loop.Info.Offset && status.SizeLimit == loop.Info.SizeLimit {
-				// keep the reference to the loop device file descriptor to
-				// be sure that the loop device won't be released between this
-				// check and the mount of the filesystem
-				sylog.Debugf("Sharing loop device %d", device)
-				loop.fd = new(int)
-				*loop.fd = loopFd
-				return nil
-			}
+		syscall.Close(loopFd)
+	}
+	return false, nil
+}
+
+// attachLoop will find a free /dev/loopXX device, or create a new one, and attach image to it.
+// For most failures with loopN, it will try loopN+1, continuing up to loop.MaxLoopDevices.
+// If there was an EAGAIN/EBUSY error on setting loop flags this is transient, and the returned
+// errTransientAttach indicates it is likely worth trying again.
+func (loop *Device) attachLoop(image *os.File, mode int, number *int) error {
+	var path string
+	// Keep track of the last transient error we hit (if any)
+	// If we fail to find a loop device, but hit at least one transient error then it's worth trying again.
+	var transientError error
+
+	// Because we hold a lock on /dev here, avoid delayed retries inside this function,
+	// as it could impact parallel startup of many instances of Singularity or
+	// other programs.
+	fd, err := lock.Exclusive("/dev")
+	if err != nil {
+		return err
+	}
+	defer lock.Release(fd)
+
+	for device := 0; device < loop.MaxLoopDevices; device++ {
+		*number = device
+
+		// Try to open the loop device, creating the device node if needed
+		loopFd, err := openLoopDev(device, mode, true)
+		if err != nil {
+			sylog.Debugf("couldn't openLoopDev loop device %d: %v", device, err)
+		}
+
+		_, _, esys := syscall.Syscall(syscall.SYS_IOCTL, uintptr(loopFd), CmdSetFd, image.Fd())
+		// On error, we'll move on to try the next loop device
+		if esys != 0 {
 			syscall.Close(loopFd)
-		} else {
-			_, _, esys := syscall.Syscall(syscall.SYS_IOCTL, uintptr(loopFd), CmdSetFd, image.Fd())
-			if esys != 0 {
-				syscall.Close(loopFd)
+			continue
+		}
+
+		if _, _, err := syscall.Syscall(syscall.SYS_FCNTL, uintptr(loopFd), syscall.F_SETFD, syscall.FD_CLOEXEC); err != 0 {
+			return fmt.Errorf("failed to set close-on-exec on loop device %s: %s", path, err.Error())
+		}
+
+		if _, _, err := syscall.Syscall(syscall.SYS_IOCTL, uintptr(loopFd), CmdSetStatus64, uintptr(unsafe.Pointer(loop.Info))); err != 0 {
+			// If we hit an error then dissociate our image from the loop device
+			syscall.Syscall(syscall.SYS_IOCTL, uintptr(loopFd), CmdClrFd, 0)
```

Comment:
clearing the backed file for the loop device isn't going to help as this error would also be returned for the next devices leading to a "no loop devices available" error